keps/sig-apps/4026-job-creation-timestamp-annotation/README.md

```diff
@@ -0,0 +1,299 @@
+
+### Graduation Criteria
+
+The feature will be released directly in Beta state since there is no benefit in having an alpha release, since we are simply adding a new annotation so there is very little risk (unlike removing an
+existing annotation which other things may depend on, for example).
+
+#### Beta
+
+- Feature implemented behind the `CronJobsScheduledAnnotation` feature gate.
+- Unit and integration tests passing.
+
+#### GA
+
+Fix any potentially reported bugs.
+
+### Upgrade / Downgrade Strategy
+
+No changes required to existing cluster to use this feature.
+
+### Version Skew Strategy
+
+N/A. This feature doesn't require coordination between control plane components,
+the changes to each controller are self-contained.
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: JobScheduledAnnotation
+  - Components depending on the feature gate:
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. If the feature gate is disabled, the CronJob controller will not add the
+scheduled timestamp as an annotation.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The CronJob controller will begin adding the scheduled timestamp as an annotation to jobs created while the feature is enabled, and existing jobs will be unaffected.
+
+###### Are there any tests for feature enablement/disablement?
+
+We plan to add unit tests.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+It will not impact already running workloads.
+
+###### What specific metrics should inform a rollback?
+
+- Users can monitor queue related metrics (e.g., queue depth and work duration) to make sure they aren't growing.
+- For CronJobs, users can also monitor `job_creation_skew_duration_seconds`.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+The feature will be tested manually prior to beta launch.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+N/A
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Check if CronJobs have the annotation `batch.kubernetes.io/cronjob-scheduled-timestamp`.
```

Comment:
That's not something I can really do if e.g. I'm managing 10,000 clusters.

I'm wondering if we should consider exposing a metric, but maybe that would be an overkill..
@soltysh - thoughts?

Reply:
In normal cases I'd agree that metric is required, but for this particular case, I think randomly checking that value is sufficient. For monitoring purposes I'd strictly rely on pre-existing metrics which monitor both the cronjob queue and the job creation skew, which should provide sufficient signal if the controller is working as expected.

Reply:
Updated.